[Feature] Support lexical parser options

colepatrickturner opened the pull request:
This adds the ability to specify `parseOptions` through makeExecutableSchema into graphql.parse - unlocking finer control of lexical options when building a schema.

# Motivation
I'm working with tooling where I would like to be able to parse an empty schema `Query` type. As of today, it's not possible with `graphql-tools` because types must have at least one field. This is possible with `graphql` currently in master, albeit as of now those options are legacy or experimental feature flags. There are other use cases, such as `noLocation` which can be used to optimize AST building. It's nice to have these options as they are now and possibly what may be added in the future.

**Creating this pull request for consideration once the master branch and all current options are deployed to NPM.**

# What's changed
- Added `parseOptions` to `makeExecutableSchema` parameters
- Carried `parseOptions` from `makeExecutableSchema` to `graphql/parse`
- Added tests for current lexical parser options.

## Current options supported:
- noLocation
- allowLegacySDLEmptyFields (pending publish)
- allowLegacySDLImplementsInterfaces (pending publish)
- experimentalFragmentVariables (pending publish)

# Example
```js
makeExecutableSchema({
  typeDefs: `
    type RootQuery {
    }
    schema {
      query: RootQuery
    }
  `,
  resolvers: {
  },
  parseOptions: {
    allowLegacySDLEmptyFields: true
  },
});
```

# Related
https://github.com/graphql/graphql-js/pull/1171
https://github.com/graphql/graphql-js/commit/c496fb72510a8da8f55ccf4c4cc9c49cfd0b656b
https://github.com/graphql/graphql-js/commit/ce0a4b9c76423edaddc2d9c65661189d0dba77cd

# Tasks
- [X] ~~If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)~~
- [X] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [X] Make sure all tests and linter rules pass
- [X] Update CHANGELOG.md with your change

freiksenet commented:
Sorry for update spam, github was misbehaving with conflict resolution. This fails because graphql-js haven't released version with those changes yet. I'll merge it once it's released.

colepatrickturner commented:
Thanks for reviewing, I'll keep an eye out for the next release of `graphql`. Cheers!

stubailo commented:
Has the upstream stuff been published yet?

colepatrickturner commented:
Not yet, it's been over a month since last release of graphql-js

colepatrickturner commented:
Release should be coming soon:
https://github.com/graphql/graphql-js/releases/tag/v0.13.0-rc.1

- Ill add support `allowLegacySDLImplementsInterfaces` before merge.

colepatrickturner commented:
[GraphQL-JS @ 0.13.0 has released](https://github.com/graphql/graphql-js/releases/tag/v0.13.0)

I rebased my PR from master, updated the package versions, and ran tests again (passing locally ✅)

freiksenet commented:
@colepatrickturner I get typescript errors locally atm:

```
src/schemaGenerator.ts(241,7): error TS2365: Operator '===' cannot be applied to types '"OperationDefinition" | "FragmentDefinition" | "SchemaDefinition" | "ScalarTypeDefinition" | "Obj...' and '"TypeExtensionDefinition"'.
```

freiksenet commented:
Other stuff this PR needs:

* [ ] Get new TS definitions that include parseOptions and update them in devDependencies
* [ ] Update `graphql` version in peerDependencies and devDependencies
* [ ] Add 0.13 to travis feature matrix
* [ ] Disable new tests for 0.11 and 0.12 using feature matrix env var
* [ ] Update tests in mergeSchema to use new inheritance syntax, put 0.12 definitions under flag.

freiksenet commented:
(Though we could possibly split the above into "support 0.13" PR)

colepatrickturner commented:
Woops I forgot to push it yesterday. Will go through the checklist

colepatrickturner commented:
- [ ] Get new TS definitions that include parseOptions and update them in devDependencies https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23525
- [X] Update graphql version in peerDependencies and devDependencies https://github.com/apollographql/graphql-tools/pull/567/commits/8e05c2a5082e7a6f47e18af594dda89535c86156
- [X] Add 0.13 to travis feature matrix https://github.com/apollographql/graphql-tools/pull/567/commits/5a93b66650ec4963f2dc3986dbfadd58c543e6b8
- [X] Disable new tests for 0.11 and 0.12 using feature matrix env var
- [ ]  Update tests in mergeSchema to use new inheritance syntax, put 0.12 definitions under flag.

freiksenet commented:
@colepatrickturner Thank you so much for handling this! Waiting for the TS to merge and I'll release.

freiksenet commented:
@colepatrickturner Could you enable me to push to your branch? I have fixes that make tests pass.

colepatrickturner commented:
@freiksenet I have added you as a contributor to my fork

freiksenet commented:
Urgh, there is issue with graphql-subscriptions not supporting 0.13 yet, so non-v5 npm fails. Waiting for this to land. https://github.com/apollographql/graphql-subscriptions/pull/135

freiksenet commented:
(╯°□°）╯︵ ┻━┻

Now it works.
